File: python/crypto/key_exchange.py
"""
Key exchange for Mind Protocol.

Sealed-box encryption for wrapping/unwrapping space keys per actor.
Uses PyNaCl's SealedBox (X25519 + XSalsa20-Poly1305) for anonymous
public-key encryption. Byte-compatible with JS libsodium sealed boxes.

Co-Authored-By: Tomaso Nervo (@nervo) <nervo@mindprotocol.ai>
"""

import base64

from nacl.public import PublicKey, PrivateKey, SealedBox


def encrypt_space_key_for_actor(space_key: bytes, actor_public_key: bytes) -> str:
    """Encrypt a space key so only the target actor can decrypt it.

    Uses a sealed box: the sender is anonymous and only the holder of the
    corresponding private key can open the box.

    Args:
        space_key: The raw space key bytes (typically 32 bytes for AES-256).
        actor_public_key: The actor's 32-byte X25519 public key.

    Returns:
        A base64-encoded string containing the sealed box ciphertext.

    Raises:
        ValueError: If the public key is not 32 bytes.
    """
    if len(actor_public_key) != 32:
        raise ValueError(
            f"Actor public key must be 32 bytes, got {len(actor_public_key)}"
        )

    pub = PublicKey(actor_public_key)
    sealed_box = SealedBox(pub)
    encrypted = sealed_box.encrypt(space_key)

    return base64.b64encode(encrypted).decode("ascii")


def decrypt_space_key_for_actor(
    encrypted_key: str,
    actor_public_key: bytes,
    actor_private_key: bytes,
) -> bytes:
    """Decrypt a sealed-box-wrapped space key using the actor's key pair.

    Args:
        encrypted_key: Base64-encoded sealed box ciphertext.
        actor_public_key: The actor's 32-byte X25519 public key.
        actor_private_key: The actor's 32-byte X25519 private key.

    Returns:
        The decrypted space key bytes.

    Raises:
        ValueError: If key lengths are incorrect.
        nacl.exceptions.CryptoError: If decryption fails.
    """
    if len(actor_public_key) != 32:
        raise ValueError(
            f"Actor public key must be 32 bytes, got {len(actor_public_key)}"
        )
    if len(actor_private_key) != 32:
        raise ValueError(
            f"Actor private key must be 32 bytes, got {len(actor_private_key)}"
        )

    priv = PrivateKey(actor_private_key)
    sealed_box = SealedBox(priv)

    ciphertext = base64.b64decode(encrypted_key)
    return sealed_box.decrypt(ciphertext)

File: lib/crypto/key_exchange.js
/**
 * Key Exchange — Sealed-box encryption for distributing space keys to actors.
 *
 * Uses libsodium sealed boxes (crypto_box_seal / crypto_box_seal_open).
 * Sealed boxes provide anonymous sender encryption — only the recipient
 * can decrypt, and the sender's identity is not revealed.
 *
 * Co-Authored-By: Tomaso Nervo (@nervo) <nervo@mindprotocol.ai>
 */

'use strict';

const sodium = require('libsodium-wrappers');

// --- Functions ---

/**
 * Encrypt a space key for a specific actor using their public key.
 *
 * Uses a sealed box — the sender is anonymous. Output is a single
 * opaque base64 string (NOT nonce:ciphertext).
 *
 * @param {Buffer} spaceKey - 32-byte space key to wrap
 * @param {Buffer} actorPublicKey - 32-byte X25519 public key of the recipient
 * @returns {Promise<string>} Base64-encoded sealed box
 */
async function encryptSpaceKeyForActor(spaceKey, actorPublicKey) {
  await sodium.ready;

  if (!Buffer.isBuffer(spaceKey) || spaceKey.length !== 32) {
    throw new Error('spaceKey must be a 32-byte Buffer');
  }
  if (!Buffer.isBuffer(actorPublicKey) || actorPublicKey.length !== 32) {
    throw new Error('actorPublicKey must be a 32-byte Buffer');
  }

  const sealed = sodium.crypto_box_seal(
    new Uint8Array(spaceKey),
    new Uint8Array(actorPublicKey),
  );

  return Buffer.from(sealed).toString('base64');
}

/**
 * Decrypt a sealed-box-encrypted space key using the actor's key pair.
 *
 * Requires BOTH public and private key — libsodium's crypto_box_seal_open
 * needs the full key pair for decryption.
 *
 * @param {string} encryptedKey - Base64-encoded sealed box
 * @param {Buffer} actorPublicKey - 32-byte X25519 public key
 * @param {Buffer} actorPrivateKey - 32-byte X25519 private key
 * @returns {Promise<Buffer>} 32-byte space key
 */
async function decryptSpaceKeyForActor(encryptedKey, actorPublicKey, actorPrivateKey) {
  await sodium.ready;

  if (typeof encryptedKey !== 'string') {
    throw new Error('encryptedKey must be a base64 string');
  }
  if (!Buffer.isBuffer(actorPublicKey) || actorPublicKey.length !== 32) {
    throw new Error('actorPublicKey must be a 32-byte Buffer');
  }
  if (!Buffer.isBuffer(actorPrivateKey) || actorPrivateKey.length !== 32) {
    throw new Error('actorPrivateKey must be a 32-byte Buffer');
  }

  const sealed = new Uint8Array(Buffer.from(encryptedKey, 'base64'));

  const opened = sodium.crypto_box_seal_open(
    sealed,
    new Uint8Array(actorPublicKey),
    new Uint8Array(actorPrivateKey),
  );

  if (!opened) {
    throw new Error('Failed to decrypt space key — invalid key pair or corrupted ciphertext');
  }

  return Buffer.from(opened);
}

// --- Exports ---

module.exports = {
  encryptSpaceKeyForActor,
  decryptSpaceKeyForActor,
};
